llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,322 @@
+      // If a page is invalid, we skip it.
+      if (PV == Other.PV)
+        return ElementIdx - Other.ElementIdx;
+
+      std::size_t ElementMin = std::min(ElementIdx, Other.ElementIdx);
+      std::size_t ElementMax = std::max(ElementIdx, Other.ElementIdx);
+      std::size_t PageMin = ElementMin / PAGE_SIZE;
+      std::size_t PageMax = ElementMax / PAGE_SIZE;
+
+      std::size_t Count = 0ULL;
+      for (std::size_t PageIdx = PageMin; PageIdx < PageMax; ++PageIdx) {
+        if (PV->PageToDataIdx[PageIdx] == InvalidPage)
+          continue;
+
+        Count += PAGE_SIZE;
+      }
+      Count += ElementMax % PAGE_SIZE;
+      Count += PAGE_SIZE - ElementMin % PAGE_SIZE;
+
+      return Count;
+    }
+
+    // When dereferencing the iterator, we materialise the page if needed.
+    T const &operator*() const {
+      assert(ElementIdx < PV->Size);
+      assert(PV->PageToDataIdx[ElementIdx / PAGE_SIZE] != InvalidPage);
+      T *PagePtr =
+          reinterpret_cast<T *>(PV->PageToDataIdx[ElementIdx / PAGE_SIZE]);
+      return *((ElementIdx % PAGE_SIZE) + PagePtr);
+    }
+
+    // Equality operator.
+    bool operator==(MaterialisedIterator const &Other) const {
+      // Iterators of two different vectors are never equal.
+      if (PV != Other.PV)
+        return false;
+      // Any iterator for an empty vector is equal to any other iterator.
+      if (PV->empty())
+        return true;
+      // Get the pages of the two iterators. If between the two pages there
+      // are no valid pages, we can condider the iterators equal.
+      std::size_t PageMin = std::min(ElementIdx, Other.ElementIdx) / PAGE_SIZE;
+      std::size_t PageMax = std::max(ElementIdx, Other.ElementIdx) / PAGE_SIZE;
+      // If the two pages are past the end, the iterators are equal.
+      if (PageMin >= PV->PageToDataIdx.size())
+        return true;
+      // If only the last page is past the end, the iterators are equal if
+      // all the pages up to the end are invalid.
+      if (PageMax >= PV->PageToDataIdx.size()) {
+        for (std::size_t PageIdx = PageMin; PageIdx < PV->PageToDataIdx.size();
+             ++PageIdx)
+          if (PV->PageToDataIdx[PageIdx] != InvalidPage)
+            return false;
+        return true;
+      }
+
+      uintptr_t Page1 = PV->PageToDataIdx[PageMin];
+      uintptr_t Page2 = PV->PageToDataIdx[PageMax];
+      if (Page1 == InvalidPage && Page2 == InvalidPage)
+        return true;
+
+      // If the two pages are the same, the iterators are equal if they point
+      // to the same element.
+      if (PageMin == PageMax)
+        return ElementIdx == Other.ElementIdx;
+
+      // If the two pages are different, the iterators are equal if all the
+      // pages between them are invalid.
+      for (std::size_t PageIdx = PageMin; PageIdx < PageMax; ++PageIdx)
+        if (PV->PageToDataIdx[PageIdx] != InvalidPage)
+          return false;
+
+      return true;
```

Comment:
You only form iterators that refer to either a materialized element or to the end, so dealing with invalid pages and the possibility that two non-equal iterators might be at the same position seems unnecessary. Also, there's no need to handle comparisons between different vectors. I think you can simplify all of this to just:
```suggestion
      return ElementIdx == Other.ElementIdx;
```

Reply:
If I drop this bit of code, I get one extra count when measuring the distance between begin and end. I need to better understand why is that...

Reply:
Have you made any progress understanding this (or perhaps the problem's gone away with some of the other changes)? I think if this simplification doesn't work, that suggests a bug somewhere else, so it'd be good to understand.

Reply:
Ping

Reply:
Sorry, I thought you noticed I changed the code as you suggested, adding asserts to make sure that iterators are indeed always valid when being compared (or == `end()`). The actual change was:

https://github.com/llvm/llvm-project/pull/66430/commits/985444327c26d5caaec416f386c41f264c08b634

This now works as expected, not sure what was the problem before, maybe I was actually constructing an invalid iterator.